testing/README.md

```diff
@@ -0,0 +1,355 @@
+
+- Use mocha’s ability to automatically finish tests when a returned promise is resolved for tests that are asynchronous and depend on promises.
+
+  ```js
+  // bad
+  test('it resolves with a value', (done) => {
+    doSomethingAsync()
+      .then((result) => {
+        assert.equal(result, 'something');
+        done();
+      })
+      .catch(() => {
+        assert(false, 'The promise was rejected :(');
+      });
+  });
+
+  // good
+  test('it resolves with a value', () => {
+    return doSomethingAsync()
+      .then((result) => {
+        assert.equal(result, 'something');
+      });
+  });
+
+  // also good, if using async/ await
+  test('it resolves with a value', async () => {
+    const result = await doSomethingAsync();
+    assert.equal(result, 'something');
+  });
+  ```
+
+[↑ scrollTo('#table-of-contents')](#table-of-contents)
+
+
+
+## Best practices
+
+- Try to limit the number of assertions made in a single test. For unit tests, you should only be testing one thing at a time (generally, an individual function, either standalone or as a method of an object), and more than one or two assertions is generally indicative that you should split your test into multiple tests.
+
+- Something something about how to organize suites (by method? by context?)
+
+- In general, each test has three steps: setup, perform the action under test, and assert based on the result.
+
+- Don't rely too heavily on mocking — maybe dependency injection instead?
+
+- Avoid using mocks.
+
+  > Why? They reverse the typical order of tests (setup, do something, assertion), which makes them harder to read in the context of other tests. Generally, you can achieve the same thing with stubs instead.
+
+[↑ scrollTo('#table-of-contents')](#table-of-contents)
+
+
+
+## Approach and types of testing
+
+Like all software testing approaches, JavaScript testing can be divided into three types: unit testing, integration testing, and functional testing.
+
+**Unit Testing** Unit testing checks whether a particular piece (unit/module) of code is working fine. The ‘unit’ is often a function (either as a method or a standalone function).
+
+**Integration Testing** Integration testing is performed when various units are integrated with each other to form a sub-system or a system. This mostly focuses in the design and construction of the software architecture.
+
+**Functional Testing** The software is tested for the functional requirements. This checks whether the application is behaving according to the specification.
+
+We currently have no recommended approach to integration or functional testing. But here are some tips to keep in mind:
+
+* Avoid assertions on CSS selectors as much as possible. These types of tests are often too brittle.
+* Keep interaction elements like "click button" as open to change as possible.
+* Test more complex flows, for example:
+  * "Add a product" as opposed to the specifics of automatic variant generation or publish visibility toggling.
+
```

Comment:
can we add UI testing to this list as well?

Selenium or an alternative is also a key piece to this equation IMO.